crates/noirc_evaluator/src/ssa.rs

```diff
@@ -59,6 +59,7 @@ pub(crate) fn optimize_into_acir(
             .print(print_ssa_passes, "After Simplifying:")
             // Run mem2reg before flattening to handle any promotion
             // of values that can be accessed after loop unrolling
+            // If this pass is missed slice merging will fail inside of flattening
```

Comment:
I'll add some more information as the comments makes it sounds like all cases of slice merging would fail. This pass handles slice mutations uncovered inside after loop unrolling. For example:
```
    let mut slice = [0; 2];
    if x != y {
        for i in 0..5 {
            slice = slice.push_back(i);
        }
    } else {
        slice = slice.push_back(x);
    }
    assert(slice[6] == 4);
    assert(slice.len() == 7);
```
This snippet requires the mem2reg pass before flattening, as we cannot simplify the slice push back instructions. This is the SSA after flattening: 
```
After Simplifying:
main
...
  b1():
    v37 = load v2
    v38 = load v4
    v39, v40 = call slice_push_back(v37, v38, Field 0)
    store v39 at v2
    store v40 at v4
    v42 = load v2
    v43 = load v4
    v44, v45 = call slice_push_back(v42, v43, Field 1)
    store v44 at v2
    store v45 at v4
    v47 = load v2
    v48 = load v4
    v49, v50 = call slice_push_back(v47, v48, Field 2)
    store v49 at v2
    store v50 at v4
    v53 = load v2
    v54 = load v4
    v55, v56 = call slice_push_back(v53, v54, Field 3)
    store v55 at v2
    store v56 at v4
    v58 = load v2
    v59 = load v4
    v60, v61 = call slice_push_back(v58, v59, Field 4)
    store v60 at v2
    store v61 at v4
    jmp b3()
  b3():
```
After Mem2reg:
```
  b1():
    v39, v40 = call slice_push_back(Field 2, [Field 0, Field 0], Field 0)
    store v39 at v2
    store v40 at v4
    v44, v45 = call slice_push_back(v39, v40, Field 1)
    store v44 at v2
    store v45 at v4
    v49, v50 = call slice_push_back(v44, v45, Field 2)
    store v49 at v2
    store v50 at v4
    v55, v56 = call slice_push_back(v49, v50, Field 3)
    store v55 at v2
    store v56 at v4
    v60, v61 = call slice_push_back(v55, v56, Field 4)
    store v60 at v2
    store v61 at v4
    jmp b3()
  b3():
```